### PR TITLE
[CS2113-W14-4] Pharmacy Inventory & Logistics Ledger (PILL)

### PR DESCRIPTION
**Pull request by yijiano**
Pharmacy Inventory & Logistics Ledger (PILL) is a Command Line Interface (CLI) tool designed to assist in
managing and tracking medicinal inventory.

### CODE REVIEW

File: src/main/java/seedu/pill/command/HelpCommand.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    /**
+     * Calls the appropriate method depending on what the user requests help for.
+     * @param command - optional user input that determines which help information is displayed to the user.
+     */
+    private void showSpecificHelp(String command) {
+        assert command != null : "Command cannot be null";
+        logger.info("Showing specific help for command: " + command);
+
+        switch (command.toLowerCase()) {
+        case "help":
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/command/HelpCommand.java]: Maybe you can refactor the magic string?

File: src/main/java/seedu/pill/util/StringMatcher.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        int[][] dp = new int[s1.length() + 1][s2.length() + 1];
+
+        for (int i = 0; i <= s1.length(); i++) {
+            for (int j = 0; j <= s2.length(); j++) {
+                if (i == 0) {
+                    dp[i][j] = j;
+                } else if (j == 0) {
+                    dp[i][j] = i;
+                } else {
+                    dp[i][j] = min(dp[i - 1][j - 1]
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/StringMatcher.java]: Maybe you can look to refactor this line to make it more readable?

File: src/main/java/seedu/pill/util/StringMatcher.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     * Finds the closest match to the input string from a list of valid strings.
+     * @param input The input string
+     * @param validStrings List of valid strings
+     * @return The closest match, or null if no close match found
+     */
+    public static String findClosestMatch(String input, List<String> validStrings) {
+        String closestMatch = null;
+        int minDistance = Integer.MAX_VALUE;
+
+        for (String valid : validStrings) {
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/StringMatcher.java:78]: valid might be a bit ambiguous, maybe you can use `validString` instead?

File: src/main/java/seedu/pill/util/Storage.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * Returns data in current line as an Item
+     * @param line Next line string read by the scanner
+     * @return The item present in the line
+     * @throws PillException if format of saved data is incorrect
+     */
+    public Item loadLine(String line) throws PillException {
+        Item item;
+        String[] data = line.split(SEPARATOR);
+        if (data.length == 3) {
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/Storage.java]: Maybe you can refactor the magic literal to make it more readable

File: src/main/java/seedu/pill/util/Storage.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            try {
+                item = new Item(data[0], Integer.parseInt(data[1]), LocalDate.parse(data[2]));
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/Storage.java]: Maybe you could refactor `data[0]` and so on such that it is more understandable what each variable means before the if statement

File: src/main/java/seedu/pill/util/Storage.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+                    if (item.getExpiryDate().isPresent()) {
+                        fw.write(SEPARATOR + item.getExpiryDate().get().toString());
+                    } else {
+                        fw.write(SEPARATOR + "");
+                    }
+
+                    fw.write(System.lineSeparator());
+                }
+            }
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/Storage.java]: Maybe you can look at refactoring this as it is deeply nested

File: src/main/java/seedu/pill/util/Parser.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+        String itemName;
+        String quantityStr = null;
+        String expiryDateStr = null;
+
+        if (quantityIndex != null && quantityIndex == splitArguments.length - 1) {
+            quantityStr = splitArguments[quantityIndex];
+            expiryDateStr = null;
+            itemName = buildItemName(splitArguments, 0, quantityIndex);
+        } else if (quantityIndex != null && quantityIndex == splitArguments.length - 2
+                && isValidDate(splitArguments[quantityIndex + 1])) {
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/Parser.java:565]: Maybe you can refactor the conditional statement out to avoid complicated espressions

File: src/main/java/seedu/pill/util/Parser.java [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+     * Only one date and one quantity are allowed.
+     *
+     * The method loops through the input array to determine the item name, quantity, and expiry date,
+     * applying default values when necessary (e.g., quantity defaults to 1 if not specified).
+     *
+     * @param arguments A string representing the user's input for adding an item
+     * @return An {@code AddItemCommand} containing the parsed item name, quantity, and optional expiry date.
+     * @throws PillException If the input format is invalid
+     */
+    private AddItemCommand parseAddItemCommand(String arguments) throws PillException {
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/Parser.java:527]: Maybe you can split this method up to multiple smaller methods instead of it all being in this long method

File: src/main/java/seedu/pill/util/ItemMap.java [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+            for (Item item : itemSet) {
+                if (item.getExpiryDate().equals(expiryDate)) {
+                    int newQuantity = item.getQuantity() + quantity;
+                    item.setQuantity(newQuantity);
+                    itemUpdated = true;
+                    expiryDate.ifPresentOrElse(
+                            expiry -> {
+                                LOGGER.info("Updated existing item with expiry date: " + newItem);
+                                System.out.println("Item already exists with the same expiry date. Updated quantity: \n"
+                                        + newItem);
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/ItemMap.java]: Maybe you can avoid lambda expressions to make it more readable for readers

File: src/main/java/seedu/pill/util/ItemMap.java [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+                    items.remove(name);
+                }
+            } else {
+                LOGGER.warning("Attempt to delete non-existent item: " + removedItem);
+                System.out.println("Item not found: " + removedItem);
+            }
+        } else {
+            LOGGER.warning("Attempt to delete non-existent item: " + name);
+            System.out.println("Item not found: " + name);
+        }
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/ItemMap.java:210]: Maybe you can make the [happy path more obvious](https://nus-cs2113-ay2425s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent) by switching this into guard clauses instead?

File: src/main/java/seedu/pill/util/ItemMap.java [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+    public Iterator<Map.Entry<String, TreeSet<Item>>> iterator() {
+        return items.entrySet().iterator();
+    }
+
+    /**
+     * Adds a new item to the list.
+     *
+     * @param newItem The item to be added.
+     */
+    public void addItem(Item newItem) {
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/ItemMap.java:73]: Maybe can look into splitting this method into several smaller methods as it is currently quite long

File: src/main/java/seedu/pill/util/ItemMap.java [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+            }
+
+            if (threshold < 0) {
+                throw new PillException(ExceptionMessages.INVALID_QUANTITY);
+            }
+
+            List<Item> filteredItems = items.values().stream()
+                    .flatMap(TreeSet::stream)
+                    .filter(item -> item.getQuantity() <= threshold)
+                    .toList();
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/ItemMap.java:337]: Maybe you can look into refactoring this to make it more readable and less complicated?

File: src/main/java/seedu/pill/util/Parser.java [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    public void parseCommand(String input) {
+        try {
+            String[] splitInput = input.split("\\s+");
+            String commandString = splitInput[0].toLowerCase();
+            String argument = splitInput.length > 1 ? splitInput[1] : null;
+            String flagStr = splitInput.length > 2 ? splitInput[2] : "";
+            String arguments = String.join(" ", Arrays.copyOfRange(splitInput, 1, splitInput.length));
+
+            switch (commandString) {
+            case "exit":
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/util/Parser.java:72]: Maybe you can look into refactoring these magic strings?

File: src/main/java/seedu/pill/command/DeleteItemCommand.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public DeleteItemCommand(String itemName, LocalDate expiryDate) {
+        this.itemName = itemName;
+        this.expiryDate = expiryDate;
+    }
+
+    @Override
+    public void execute(ItemMap itemMap, Storage storage) throws PillException {
+        // Looks stupid, but this way I don't handle Optionals in this class
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/command/DeleteItemCommand.java:26]: Maybe you can rephrase this to [write comments targeting other programmers reading the code](https://nus-cs2113-ay2425s1.github.io/website/se-book-adapted/chapters/codeQuality.html#write-to-the-reader)

File: src/main/java/seedu/pill/command/HelpCommand.java [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+            showExitHelp();
+            break;
+        default:
+            String closestMatch = StringMatcher.findClosestMatch(command, VALID_COMMANDS);
+            if (closestMatch != null) {
+                System.out.println("Did you mean: " + closestMatch + "?");
+                showSpecificHelp(closestMatch);
+            } else {
+                suggestSimilarCommand(command);
+            }
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/command/HelpCommand.java:220]: Maybe this can be changed to a guard clause instead to make the happy path more prominent

File: src/main/java/seedu/pill/command/HelpCommand.java [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+ */
+public class HelpCommand extends Command {
+    private static final Logger logger = PillLogger.getLogger();
+    private static final List<String> VALID_COMMANDS = Arrays.asList(
+            "help", "add", "delete", "edit",
+            "list", "exit", "stock-check");
+    private final String commandName;
+    private final boolean verbose;
+
+    public HelpCommand(String commandInput, boolean verbose) {
```
Comment (thomasjlalba) [inline src/main/java/seedu/pill/command/HelpCommand.java:38]: Maybe you can change the boolean variable name to make it more obvious it is a boolean

**Review by thomasjlalba — COMMENTED** [review]
Do remember to ensure your code is more readable for others and to simplify expressions. Keep up the good work

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,38 +1,252 @@
...
+
+<!-- @@author cnivedit -->
+
+#### AddItemCommand
+
+<img src = "diagrams/AddItemCommand-SequenceDiagram.png"/>
+
+<!-- @@author yakultbottle -->
+
+### Storage
```
Comment (Bev-low) [inline docs/DeveloperGuide.md:113]: I think these classes can have UML diagrams

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,38 +1,252 @@
...
+### Commands
+
+Each user action (e.g. adding, deleting, or editing an item) is mapped to a specific command class.
+These classes handle the logic for interpreting the input and calling the appropriate methods from the ItemMap and
+Storage classes.
+
+Example:
+
+```
+AddItemCommand command = new AddItemCommand(itemName, quantity, expiryDate);
```
Comment (Bev-low) [inline docs/DeveloperGuide.md:94]: I think this can be put in the implementation section, following the example they give us.

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,38 +1,252 @@
...
+ItemMap. While it has other dependencies, such as File and FileWriter from
+the Java standard library, PillException is the only custom class it depends on.
+
+<!-- @@author cnivedit -->
+
+### Logging
 
+**API**: PillLogger.java
+
+<img src = "diagrams/PillLogger.png" alt="Component Diagram for PillLogger"/>
```
Comment (Bev-low) [inline docs/DeveloperGuide.md:289]: This diagram looks like it doesnt follow the standard

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1,38 +1,252 @@
...
+![](diagrams/Item-ClassDiagram.png)
+
+Quantity will always be a positive integer, and if no quantity is specified
+in the constructor, the default value is 1. Similarly, if no value is provided
+for expiry date, then it will be an Optional.empty().
+
+The ItemMap class contains a key-value pair, implemented as a Map, from the
+item name(String) to the item(TreeSet\<Item>)
+
+![](diagrams/ItemMap-ClassDiagram.png)
```
Comment (Bev-low) [inline docs/DeveloperGuide.md]: I dont think this follows the standard also, I am not sure what the dotted box "item" mean on comparable

Comment (Bev-low) [reply docs/DeveloperGuide.md]: should do comparable:item

**Review by Bev-low — COMMENTED** [review]
Overall looks neat

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1,38 +1,252 @@
...
+instance. Classes within the application use `getLogger()` to record events, without needing to set up or manage their own
+loggers.
+
+<!-- @@author yakultbottle -->
+
+### Item and ItemMap
+
+The Item class has three private variables, a name, a quantity, and an
+expiry date. An Item may or may not have an expiry date, so we store it
+as an Optional, which handles empty values for us without using null.
```
Comment (YukeeHong) [inline docs/DeveloperGuide.md]: could briefly explain why Optional was chosen over null

File: docs/DeveloperGuide.md (hunk 19 above)
Comment (YukeeHong) [reply docs/DeveloperGuide.md:289]: yeah, i think missing some tag and some place should use dotted lines

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -1,38 +1,252 @@
...
+| v2.0    | logistics in-charge | list all expiring items                   | remember to clear them out                                   |
+| v2.0    | logistics in-charge | list all items that need to be restocked  | place orders before items run out                            |
 
 ## Non-Functional Requirements
 
-{Give non-functional requirements}
+* Technical Requirements: Any mainstream OS, i.e. Windows, macOS or Linux, with Java 11 installed. Instructions for downloading Java 11 can be found [here](https://www.oracle.com/sg/java/technologies/javase/jdk11-archive-downloads.html).
+* Project Scope Constraints: The application should only be used for tracking. It is not meant to be involved in any form of monetary transaction.
+* Project Scope Constraints: Data storage is only to be performed locally.
+* Quality Requirements: The application should be able to be used effectively by a novice with little experience with CLIs.
```
Comment (YukeeHong) [inline docs/DeveloperGuide.md]: can be more detailed and concrete

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -1,38 +1,252 @@
...
+```
+AddItemCommand command = new AddItemCommand(itemName, quantity, expiryDate);
+command.execute(itemMap, storage);
+```
+
+<!-- @@author cnivedit -->
+
+#### AddItemCommand
+
+<img src = "diagrams/AddItemCommand-SequenceDiagram.png"/>
```
Comment (YukeeHong) [inline docs/DeveloperGuide.md]: some part like print  and log can be simplified

Comment (DanLinhHuynh-Niwashi) [inline docs/DeveloperGuide.md]: Would it be better to add some activation bars to show if the objects are working?

File: docs/DeveloperGuide.md [hunk 24]
```diff
@@ -1,38 +1,252 @@
...
+10. **ItemMap**: A class to store items in a map-like structure.
+11. **Command**: A class to represent a command that can be executed by the application.
+12. **Parser**: A class to parse user input and return the corresponding command.
+13. **Pill**: The main class that initializes the application and starts the CLI.
+
+### High-Level Overview
+
+The high-level overview of the project structure is as follows:
+
+<img src = "diagrams/High-Level-Overview.png" alt="High Level Overview of PILL"/>
```
Comment (DanLinhHuynh-Niwashi) [inline docs/DeveloperGuide.md:62]: You may try to consider the direction of arrows. Since it's is MVC model, the arrow from View to Controller might be a two way arrow (e.g. View send request to Controller and Controller response with data?), the same for Controller to Model components

File: docs/DeveloperGuide.md [hunk 25]
```diff
@@ -1,38 +1,252 @@
...
+<!-- @@author yijiano -->
+
+PILL uses the following tools for development:
+1. [JUnit 5](https://junit.org/junit5/) - Used for testing.
+2. [Gradle](https://gradle.org/) - Used for build automation.
 
 ## Design & implementation
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+The project is designed using the Model-View-Controller (MVC) architecture, with the following components:
```
Comment (DanLinhHuynh-Niwashi) [inline docs/DeveloperGuide.md:41]: Since Model - Controller - View is at a higher level comparing to other classes (e.g. Model contains Item and ItemClass) and they are not really a class as well, you may try to create a hierarchy, just to distinguish between the architecture and the actual implementation. For example:
3. Controller:
3.1. Item
3.2. ItemMap

File: docs/DeveloperGuide.md [hunk 26]
```diff
@@ -1,38 +1,252 @@
...
+<!-- @@author cxc0418 -->
+
+### UI and I/O
+
+The program uses a command-line interface (CLI) for interaction with the user.
+It receives input as text commands, processes these commands, and provides feedback through the console.
+
+The UI consists of the following components:
+
+1. **Parser**: The Parser class is responsible for interpreting user input and returning the corresponding command.
```
Comment (DanLinhHuynh-Niwashi) [inline docs/DeveloperGuide.md:74]: As I think UI is the 'View' part in your architecture, should it be better to make this parser a part of the Controller instead of View? As the Controller in MVC is responsible for handling user input, processing it, and coordinating between the Model and the View, I think the parser is nearer to a Controller component than a View component

File: docs/DeveloperGuide.md [hunk 27]
```diff
@@ -1,38 +1,252 @@
...
+### High-Level Overview
+
+The high-level overview of the project structure is as follows:
+
+<img src = "diagrams/High-Level-Overview.png" alt="High Level Overview of PILL"/>
+
+
+<!-- @@author cxc0418 -->
+
+### UI and I/O
```
Comment (DanLinhHuynh-Niwashi) [inline docs/DeveloperGuide.md:67]: Will it be good to put these implementation into a hierarchy corresponding to your architecture? To make a clearer view of how each component related to the high-level architecture!

File: docs/diagrams/High-Level-Overview.puml
Comment (Zackermax) [inline docs/diagrams/High-Level-Overview.puml:1]: The diagram should not have class symbols (green C) and the arrows should be orthogonal.

File: docs/diagrams/AddItemCommand-SequenceDiagram.puml
Comment (Zackermax) [inline docs/diagrams/AddItemCommand-SequenceDiagram.puml:1]: The diagram is missing activation boxes

File: docs/diagrams/PillLogger.puml
Comment (Zackermax) [inline docs/diagrams/PillLogger.puml:1]: The arrows should be orthogonal.

File: docs/DeveloperGuide.md [hunk 28]
```diff
@@ -1,38 +1,252 @@
...
+
+<!-- @@author yakultbottle -->
+
+### Item and ItemMap
+
+The Item class has three private variables, a name, a quantity, and an
+expiry date. An Item may or may not have an expiry date, so we store it
+as an Optional, which handles empty values for us without using null. 
+
+![](diagrams/Item-ClassDiagram.png)
```
Comment (LTK-1606) [inline docs/DeveloperGuide.md]: Are your notations used correctly? Should it be "function(): 'return type'"?

File: docs/diagrams/AddItemCommand-SequenceDiagram.puml
Comment (Zackermax) [inline docs/diagrams/AddItemCommand-SequenceDiagram.puml:1]: The formatting for the "else" condition in the second section is not formatted properly.

File: docs/DeveloperGuide.md (hunk 23 above)
Comment (LTK-1606) [inline docs/DeveloperGuide.md]: Are these notations correct? Should it include activation bars?

File: docs/DeveloperGuide.md (hunk 27 above)
Comment (LTK-1606) [inline docs/DeveloperGuide.md:67]: Should there be more visuals and diagrams to make this clearer?

File: docs/DeveloperGuide.md [hunk 29]
```diff
@@ -1,38 +1,252 @@
...
+- [Developer Guide](#developer-guide)
+    - [Table of Contents](#table-of-contents)
+    - [Acknowledgements](#acknowledgements)
+    - [Design \& Implementation](#design--implementation)
+        - [UI and I/O](#ui-and-io)
+        - [Commands](#commands)
+        - [Storage](#storage)
+        - [Item and ItemMap](#item-and-itemmap)
+        - [DateTime](#datetime)
+        - [Exceptions and Logging](#exceptions-and-logging)
```
Comment (LTK-1606) [inline docs/DeveloperGuide.md]: Are these links supposed to be here? Should they be broken?

**Review by LTK-1606 — COMMENTED** [review]
Looks neat, but notations used may not follow requirements, some broken links and maybe more diagrams would be better
